### PR TITLE
Siamese Network s43032053

### PR DESCRIPTION
**Pull request by richardchantra**
Dear COMP3710 teaching team,

Name: Richard Chantra
Student Number: s43032053
Project: #6 Siamese Network

I have 5 files for this project:
modules.py: Defines Siamese network, MLP classifier, loss functions, and evaluation
dataset.py: Manages data loading, augmentation, and data balancing
train.py: Trains the Siamese Network and MLP classifier
README.md: Project documentation

Thank you

### CODE REVIEW

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point

**Difficulty**: Hard

**Readme**: Excellent
The flow of information is clear and logical.
The algorithm and usage are well-explained.
Relevant plots (contrastive and classification losses, 2D manifolds, confusion matrix, ROC, and Siamese architecture) are provided, enhancing the explanation.

**Commit Messages**: Very good and detailed, providing clarity on the development process.

**Code**
Uses a contrastive loss network (two identical networks).
Well-designed with clear and sufficient comments throughout the code.

**Functionality/Performance**
The Siamese network has converged, but the resulting manifold is not well-separated.
The classifier performs well in terms of overall accuracy, but:
_Sensitivity_: Low at 32%, indicating poor performance in identifying malignant cases.
_Recall_: Excellent for benign cases at 99%, but only 1% for malignant cases.
Different rebalancing strategies have been adopted, but further experimentation is needed to achieve balanced performance, particularly to improve the model's ability to classify malignant cases.

**General Comments**
This pull request is targeting the main branch.
The solution is appropriate for the problem but suffers from poor performance due to class imbalance and possibly an inadequate feature-extraction backbone.
A pre-trained ResNet is used for feature extraction.
A significant factor affecting the separation between classes might be the feature-extraction backbone’s strength and the chosen data balancing method.

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **20**

Solves problem: The solution is well-suited to the problem (**5**). 
_The training manifold is effectively separated, and the classifier shows strong performance, as reflected by the accuracy and reported confusion matrix. 
Triplet loss is utilized, which is suitable for the problem.
Class imbalance is addressed through undersampling of benign cases._

Implementation functions as intended  (**2**)

Good design: Well-designed (**1**)

Commenting: Clear and sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

Note:
1. Well understaning of the medical images challanges and the triplet loss.
2. Future suggestions are well crafted and might improve the performance.
3. the specificity (76.07%) and sensitivity (88.03%) are almost balanced which means the model is able to differentiate between the two classess efficiently

_Using an algorithmic approach to undersample the majority class is indeed recommended over random sampling, especially in the context of medical imaging_ to maintain informative samples, improve generalization, and avoide loss of critical information.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Model affected by imbalance.</td><td>-3</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Has data splitting. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. </td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>Feedback action required: Feedback marks possible +2 if the requested changes are made. Merge conflicts due to changes to repo README files.</td><td>-2</td></tr>
	<tr><td></td><td>Request Description is adequate. </td><td>-1</td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-7</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by richardchantra** [comment]
Done, thanks team!

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files (LICENSE). Please update for merge, doesn't affect grade.
